src/backend/cdb/cdbpartition.c

```
@@ -4172,35 +4158,49 @@ selectListPartition(PartitionNode *partnode, Datum *values, bool *isnull,
+ * cdb_reterieve_btree_op
+ *  Retrieves btree operator according to the opclass, operands types and strategy number.
  *
+ *  Input:
+ *   opclass: The opclass is used when defining the partition table.
+ *   lhstypid: The left hand side operand type oid.
+ *   rhstypid: The right hand side operand type oid.
+ *   strategy: The strategy number of the operator.
  */
 static Oid
-get_less_than_oper(Oid lhstypid, Oid rhstypid, bool strictlyless)
+cdb_retrieve_btree_op(Oid opclass, Oid lhstypid, Oid rhstypid, int16 strategy)
 {
-	Value	   *str = strictlyless ? makeString("<") : makeString("<=");
-	Value	   *pub = makeString("pg_catalog");
-	List	   *opname = list_make2(pub, str);
-
-	Oid			funcid = get_opfuncid_by_opname(opname, lhstypid, rhstypid);
+	Oid opfamily = get_opclass_family(opclass);
+	Oid opno = InvalidOid;
+	/*
+	 * There's no associated operator for domain types, we need to fetch the base type
+	 * of these types.
+	 */
+	Oid lbasetypid = OidIsValid(lhstypid) ? getBaseType(lhstypid) : lhstypid;
+	Oid rbasetypid = OidIsValid(rhstypid) ? getBaseType(rhstypid) : rhstypid;
 
-	list_free_deep(opname);
+	opno = get_opfamily_member(opfamily, lbasetypid, rbasetypid, strategy);
+	if (!OidIsValid(opno))
+	{
+		/*
+		 * There're cases that we cannot fetch operator based on types of the operands.
+		 * e.g.,
+		 *   CREATE TABLE part_tbl_varchar (a VARCHAR(15) NOT NULL, b VARCHAR(8) NOT NULL)
+		 *     DISTRIBUTED BY (a)
+		 *     PARTITION BY RANGE(b)
+		 *       (START('v1') END('v5'), START('v5') END('v9'), DEFAULT PARTITION def);
+		 * The recorded opclass for the partitioning key is text_ops (3126), though the column
+		 * type is 'VARCHAR'. We use IsBinaryCoercible to handle these binary-compatible types.
+		 */
+		Oid inctypid = get_opclass_input_type(opclass);
+		if (IsBinaryCoercible(lbasetypid, inctypid) && IsBinaryCoercible(rbasetypid, inctypid))
+			opno = get_opfamily_member(opfamily, inctypid, inctypid, strategy);
```

if we still cannot find any, maybe throw an error?